src/plugins/content_management/server/event_stream/es/es_event_stream_client.ts

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type { estypes } from '@elastic/elasticsearch';
+import { KueryNode, nodeBuilder, toElasticsearchQuery } from '@kbn/es-query';
+import type { EsClient, EsEventStreamEventDto } from './types';
+import type {
+  EventStreamClient,
+  EventStreamClientFilterOptions,
+  EventStreamClientFilterResult,
+  EventStreamEvent,
+  EventStreamLogger,
+} from '../types';
+import { EsEventStreamNames } from './es_event_stream_names';
+import { EsEventStreamInitializer } from './init/es_event_stream_initializer';
+import { eventToDto, dtoToEvent } from './util';
+
+export interface EsEventStreamClientDependencies {
+  baseName: string;
+  kibanaVersion: string;
+  logger: EventStreamLogger;
+  esClient: Promise<EsClient>;
+}
+
+const sort: estypes.Sort = [
+  {
+    // By default we always sort by event timestamp descending.
+    '@timestamp': {
+      order: 'desc',
+    },
+
+    // Tie breakers for events with the same timestamp.
+    subjectId: {
+      order: 'desc',
+    },
+    objectId: {
+      order: 'desc',
+    },
+    predicate: {
+      order: 'desc',
+    },
+  },
+];
+
+export class EsEventStreamClient implements EventStreamClient {
+  readonly #names: EsEventStreamNames;
+
+  constructor(private readonly deps: EsEventStreamClientDependencies) {
+    this.#names = new EsEventStreamNames(deps.baseName);
+  }
+
+  public async initialize(): Promise<void> {
+    const initializer = new EsEventStreamInitializer({
+      names: this.#names,
+      kibanaVersion: this.deps.kibanaVersion,
+      logger: this.deps.logger,
+      esClient: this.deps.esClient,
+    });
+    await initializer.initialize();
+  }
+
+  public async writeEvents(events: EventStreamEvent[]): Promise<void> {
+    const esClient = await this.deps.esClient;
+    const operations: Array<estypes.BulkOperationContainer | EsEventStreamEventDto> = [];
+
+    for (const event of events) {
+      const dto = eventToDto(event);
+
+      operations.push({ create: {} }, dto);
+    }
+
+    const { errors } = await esClient.bulk(
+      {
+        index: this.#names.dataStream,
+        operations,
+      },
+      {
+        maxRetries: 0,
```

Comment:
Yes, the reason is that our requests (and many other requests Kibana does, btw) are not idempotent. So, it could result in doubly, triply, ... logging the same event.

Reply:
We had an idea to fix this in the event log, but I don't think we implemented it.  Generate the doc ids manually, when added to the queue.  Then you can write them as many times as you want, but there will only be one document!

Reply:
Yes, we wanted to do the same, but this fix was faster :)